### ISSUE TITLE
PyMuPDF 1.13.3 corrupts multibyte text in content streams

### ISSUE THREAD

**Issue by jbarlow83** [rk700/PyMuPDF#170]
The attached file was produced by using Tesseract 4.0 to produce an OCR PDF of a test image.
[linn.pdf](https://github.com/rk700/PyMuPDF/files/1993042/linn.pdf)

After opening and saving with PyMuPDF...

```python
doc = fitz.open('linn.pdf')
doc.save('linnfitz.pdf')
```

...the content stream will be corrupted, as in the output file [linnfitz.pdf](https://github.com/rk700/PyMuPDF/files/1993049/linnfitz.pdf).

It appears that PyMuPDF or MuPDF transforms text originally encoded in UTF-16BE such as

```
[ <005400680065> ] TJ
```

into

```
[(\000T\000h\000e)] TJ
```

that is, nulls are inserted between each character.

This prevents PDF viewers from interpreting the text correctly. It's not possible to search for or copy-paste text from the output file anymore.

Probably this is related to #153, but the weird part is that 1.12.5 (which includes that fix) works correctly.

**Comment by JorjMcKie** [comment]
I have tried to recreate the problem. These are my findings:

* saving the input `linn.pdf` under a new name without any options or only using `garbage` and `deflate` options did not lead to issues. See the following logs:

```python
Python 3.6.5 (v3.6.5:f59c0932b4, Mar 28 2018, 17:00:18) [MSC v.1900 64 bit (AMD64)] on win32
Type "copyright", "credits" or "license()" for more information.
>>> import fitz
>>> doc1 = fitz.open("linn.pdf")
>>> doc1[0]._getContents()
[10]
>>> content1 = doc1._getXrefStream(10)
>>> doc1.save("linn2.pdf")
>>> doc2 = fitz.open("linn2.pdf")
>>> doc2[0]._getContents()
[10]
>>> content2 = doc2._getXrefStream(10)
>>> content1 == content2
True
>>> 
```
```python
Python 3.6.5 (v3.6.5:f59c0932b4, Mar 28 2018, 17:00:18) [MSC v.1900 64 bit (AMD64)] on win32
Type "copyright", "credits" or "license()" for more information.
>>> import fitz
>>> doc1 = fitz.open("linn.pdf")
>>> content1 = doc1._getXrefStream(10)
>>> doc1.save("linn2.pdf", garbage = 4, deflate = True)
>>> doc2 = fitz.open("linn2.pdf")
>>> doc2[0]._getContents()
[10]
>>> content2 = doc2._getXrefStream(10)
>>> content1 == content2
True
>>> 
```
* However, using the `clean` option **does** lead to the issue you describe! This is not only true for PyMuPDF. Same thing happens using the respective MuPDF cli tool option: `mutool clean -c linn.pdf linn2.pdf`. The output PDF `linn2.pdf` is blank when displayed with a PDF reader and the contents object looks like you describe.

So the issue is not related to PyMuPDF (and therefore not to the resolced issue #153 you mentioned).

If you allow me, I would like to use your `linn.pdf` to submit a bug in MuPDF's bug system.

Here is a part of the "cleaned" contents object. Looks like if you replace every string `\000` by a NULL string, everything might be allright ... I will give that a try.
```
BT
3 Tr
1 0 0 1 211.68 744 Tm
/f-0-0 21 Tf
117.334 Tz
[(\000T\000h\000e)] TJ
42.96 0 Td
104.616 Tz
[(\000L\000i\000n\000n\000S\000e\000q\000u\000e\000n\000c\000e\000r)] TJ
-114.96 -20.64 Td
105.15 Tz
[(\0003\0002)] TJ
27.84 0 Td
102.862 Tz
[(\000T\000r\000a\000c\000k)] TJ
60 0 Td
133.142 Tz
[(\000M\000I\000D\000I)] TJ
62.4 0 Td
104.858 Tz
[(\000S\000e\000q\000u\000e\000n\000c\000e)] TJ
94.08 0 Td
105.144 Tz
[(\000R\000e\000c\000o\000r\000d\000e\000r)] TJ
ET
```

**Comment by jbarlow83** [comment]
I repeated your test. Our results do not agree. `clean=False` does not change anything either.

```python
In [1]: import fitz

In [2]: doc1=fitz.open('linn.pdf')

In [3]: doc1[0]._getContents()
Out[3]: [10]

In [4]: content1 = doc1._getXrefStream(10)

In [5]: doc1.save('linn2.pdf')

In [6]: doc2 = fitz.open('linn2.pdf')

In [7]: doc2[0]._getContents()
Out[7]: [10]

In [8]: content2 = doc2._getXrefStream(10)

In [9]: content1 == content2
Out[9]: False

In [10]: fitz.version
Out[10]: ('1.13.3', '1.13.0', '20180503151204')

In [12]: content2[:100]
Out[12]: b'q\n612 0 0 792 0 0 cm\n/Im1 Do\nQ\nq\nBT\n1 0 0 1 211.68 744 Tm\n1.17334 Tz\n/f-0-0 21 Tf\n3 Tr\n[(\\000T\\000h\\'

```

You can use linn.pdf for anything. It's from Wikipedia, under a Creative Commons license.

**Comment by JorjMcKie** [comment]
I have to confess: the situation completely confuses me!

1. MuPDF produces undisplayable files if "clean" option is used (`mutool clean -c file.pdf`). This is **always** the case: whether the PDF contains multi-byte characters or just simple, built-in fonts like Helvetica. I ended up submitting a bug report to them, which just included a simple PDF only containing 2 lines of simple Helvetica text. When **not** using the "clean" option, the output was ok, e.g. something like `mutool clean -ggggz file.pdf`.

2. PyMuPDF for Windows, Py3.6 / 3.7 64bit works for all save options except when "clean" is specified - this will render the PDF non-displayable.

3. PyMuPDF for Windows Py2.7 32bit **always** works, but the output is in format `clean = True` (i.e. uncompressed `/Contents` object) when no options are used! When I specify "garbage" and "deflate" everything stays ok and compression takes place as expected.

4. PyMuPDF Ubuntu Py3.5 64bit  (Windows subsystem), behaves exactly like Win Py2.7 32bit.

5. PyMuPDF Ubuntu Py3.6 64bit (VM), behaves exactly like Win Py3.6 64bit (point 2 above).

6. Your situation appears as if "clean" were specified and is wrong in the same way as point 2 with "clean" above.

I fail to see a pattern. I will now first look to get my Windows versions to a consistent behavior. I hope to see what has been going wrong. After that I will propagate the corrections to the other platforms.

**Comment by JorjMcKie** [comment]
I have the solution - MuPDF has answered very quickly:
There exist 2 independent PDF save options "clean" and "sanitize". However, they **must not be set independently** ... hm. I had been using a non-initialized save options-structure in the v1.13.3 build, which could randomly lead to those 2 options being differently set.
This obviously has happened in your case.
Now I am initializing this options structure with all zeroes, and ensure the "clean" and "sanitize" are set to the same value.
This is resulting in always-correct files for all my Windows Python versions.
I will upload a v1.13.4 during the day hoping your issue will be resolved.

**Comment by JorjMcKie** [comment]
ooops - should be close by **you**

**Comment by JorjMcKie** [comment]
In the meantime I also tested v1.13.4 successfully on my Ubuntu systems 16.x (Python 3.5) and 17.10 (Py 3.6).

**Comment by jbarlow83** [comment]
I'm on a Mac. Did you generate a wheel for that? (It's not posted in Github releases.)

I could build my own, but I think that defeats the purpose in case there's something in your process that contributes.

**Comment by JorjMcKie** [comment]
Hmpf - like ever so often, the Mac generation is causing difficulties. It is now hanging in Travis for an hour and doesn't move.
Let make one more try. I'll notify you.

**Comment by JorjMcKie** [comment]
No, sorry. The Mac job in Travis just doesn't start. I'm afraid you have to generate it yourself.

**Comment by jbarlow83** [comment]
My Mac builds aren't working on Travis today either. I think they're
underprovisioned - usually very busy in daytime US.
